### DISCUSSION TITLE
Will throttling and max_job_runtime be supported in Sidekiq::IterableJob?

### DISCUSSION THREAD

**Discussion by ricksuggs** [sidekiq/sidekiq#6432] [General]
I currently use Shopify's job-iteration, but would like to migrate to sidekiq's Sidekiq::IterableJob to reduce dependencies.
However, it looks like the [throttling module](https://github.com/fatkodima/sidekiq-iteration/blob/master/guides/throttling.md) was not ported over from sidekiq-iteration.

Another useful option is [max_job_runtime](https://github.com/fatkodima/sidekiq-iteration/blob/master/guides/best-practices.md#per-job-max-job-runtime), which makes a job interrupt itself every interval.

Are there plans to include these options?
Thank you

**Comment by mperham** [comment]
It's possible but I wanted to get the basic iteration feature released and tested widely before adding "trickier" options which might have subtle edge cases.

**Reply by rafal-brize** [reply]
Hey @mperham ,

I was wondering if you had time to give it a thought?

We would also like to use IterableJob and migrate out of Batches.

Our single iteration spawns a lot of smaller jobs, and using batches (utilising on_complete and parent batches) we are able to wait for all of them to complete in order to schedule next batch. 

Implementation using IterableJob would be much simpler, but without `max_job_runtime` am worried we would flood redis on bigger jobs - worried about the memory not the performance obviously.

**Reply by mperham** [reply]
I'm unclear why max_job_runtime is necessary for you. Could you explain more?

**Reply by rafal-brize** [reply]
Sure, i'll try my best.

Processing csv ~30k rows.
We batch the rows for faster processing, each batch takes 20s to process
Each row processing schedules additional ~15 jobs.
So during processing of a file we might schedule ~450k jobs.
This is just single file. We might have multiple files being processed in parallel which multiplies number of jobs put in redis.

Now our current approach is "batches of batches of rows" (Sidekiq Pro)
We wait for batch of rows to complete (along with all subsequent jobs) before we schedule next batch.
That limits the number of jobs in redis at a time and we can finetune it with the size of a batch.

I we were to process that with IterableJob the worker thread would process the file non stop, putting all that jobs in redis.
If we were to stop after some time and reschedule, we could use the fact that the queue is a FIFO and most of the jobs scheduled by iterations would finish before picking up next iterations.

We are also using Sidekiq Ent so i was able to do some alternative approach using limiters, but that seams like an overkill

```
  def around_iteration
    flush_state
    limiter.within_limit do
      super
    end
  end
```

I hope this explanation helps understanding what we are dealing with.
And let me know what you think, thanks!

**Comment by mperham** [comment]
I'm concerned `max_job_runtime` is arbitrary and there are alternative implementation paths available today. For example, you can implement a migration job which iterates thru 1000 elements and then schedules a job  in `on_complete` to migrate the next 1000 elements 1 minute from now. Is one easier to write vs the other? Is one easier to test?

**Reply by rafal-brize** [reply]
Its been added to sidekiq 8.0.8

https://github.com/sidekiq/sidekiq/blob/main/Changes.md#808
